cfgmgr/stpmgr.cpp

```diff
@@ -0,0 +1,1110 @@
+ * its subsidiaries.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "exec.h"
+#include "stpmgr.h"
+#include "logger.h"
+#include "tokenize.h"
+#include "warm_restart.h"
+#include "stp_ipc.h"
+
+#include <iostream>
+#include <algorithm>
+#include <sstream>
+#include <thread>
+
+#include <net/if.h>
+#include <sys/ioctl.h>
+#include <sys/stat.h>
+
+using namespace std;
+using namespace swss;
+
+StpMgr::StpMgr(DBConnector *confDb, DBConnector *applDb, DBConnector *statDb,
+        const vector<TableConnector> &tables) :
+    Orch(tables),
+    m_cfgStpGlobalTable(confDb, CFG_STP_GLOBAL_TABLE_NAME),
+    m_cfgStpVlanTable(confDb, CFG_STP_VLAN_TABLE_NAME),
+    m_cfgStpVlanPortTable(confDb, CFG_STP_VLAN_PORT_TABLE_NAME),
+    m_cfgStpPortTable(confDb, CFG_STP_PORT_TABLE_NAME),
+    m_cfgLagMemberTable(confDb, CFG_LAG_MEMBER_TABLE_NAME),
+    m_cfgVlanMemberTable(confDb, CFG_VLAN_MEMBER_TABLE_NAME),
+    m_stateVlanTable(statDb, STATE_VLAN_TABLE_NAME),
+    m_stateLagTable(statDb, STATE_LAG_TABLE_NAME),
+    m_stateStpTable(statDb, STATE_STP_TABLE_NAME),
+    m_stateVlanMemberTable(statDb, STATE_VLAN_MEMBER_TABLE_NAME)
+{
+    SWSS_LOG_ENTER();
+    l2ProtoEnabled = L2_NONE;
+
+    stpGlobalTask = stpVlanTask = stpVlanPortTask = stpPortTask = false;
+
+    // Initialize all VLANs to Invalid instance
+    fill_n(m_vlanInstMap, MAX_VLANS, INVALID_INSTANCE);
+
+    int ret = system("ebtables -D FORWARD -d 01:00:0c:cc:cc:cd -j DROP");
+    SWSS_LOG_DEBUG("ebtables ret %d", ret); 
+}
+
+void StpMgr::doTask(Consumer &consumer)
+{
+    auto table = consumer.getTableName();
+
+    SWSS_LOG_INFO("Get task from table %s", table.c_str());
+
+    if (table == CFG_STP_GLOBAL_TABLE_NAME)
+        doStpGlobalTask(consumer);
+    else if (table == CFG_STP_VLAN_TABLE_NAME)
+        doStpVlanTask(consumer);
+    else if (table == CFG_STP_VLAN_PORT_TABLE_NAME)
+        doStpVlanPortTask(consumer);
+    else if (table == CFG_STP_PORT_TABLE_NAME)
+        doStpPortTask(consumer);
+    else if (table == CFG_LAG_MEMBER_TABLE_NAME)
+        doLagMemUpdateTask(consumer);
+    else if (table == STATE_VLAN_MEMBER_TABLE_NAME)
+        doVlanMemUpdateTask(consumer);
+    else
+        SWSS_LOG_ERROR("Invalid table %s", table.c_str());
+}
+
+void StpMgr::doStpGlobalTask(Consumer &consumer)
+{
+    SWSS_LOG_ENTER();
+
+    if (stpGlobalTask == false)
+        stpGlobalTask = true;
```

Comment:
can you please explain what is the role of this variable? 

I see it being used in StpPortTask and other functions and return if this is not set. Is this to ensure that STP is enabled globally before its enabled on port  - during config load? 

If yes, then a better option would be to check the linux stp bridge state/some underlying resource which would be created in kernel and based on then decide whether to "defer & retry" in other task.

Reply:
stpGlobalTask should be done before any other stp tasks like vlan, port configurations are performed. Reason is, stpGlobalTask initializes the global stp bridge config, setsup stpmgr stack with all required bridge initialization config, so that per VLAN or Port can be added on top. Hence without this set , all other tasks are returned without action.

Reply:
You are correct; there will be no action on VLAN or port-related configurations when stpGlobalTask is not set. However, these requests will remain in the queue, and once stpGlobalTask is set, the VLAN and port-related configurations will be processed from the queue. This sequencing is necessary in the case of a device reboot with a saved configuration, as Redis-DB can be read in any order. Therefore, ensuring the correct sequence of events is essential.